crates/polars-stream/src/nodes/parquet_source.rs

```diff
@@ -0,0 +1,1628 @@
+    /// * `self.physical_predicate`
+    fn init_row_group_decoder(&self) -> RowGroupDecoder {
+        debug_assert!(
+            !self.projected_arrow_fields.is_empty()
+                || self.file_options.with_columns.as_deref() == Some(&[])
+        );
+        debug_assert_eq!(self.predicate.is_some(), self.physical_predicate.is_some());
+
+        let paths = self.paths.clone();
+        let hive_partitions = self.hive_parts.clone();
+        let include_file_paths = self.file_options.include_file_paths.clone();
+        let projected_arrow_fields = self.projected_arrow_fields.clone();
+        let row_index = self.file_options.row_index.clone();
+        let physical_predicate = self.physical_predicate.clone();
+        let cpu_runtime = self.cpu_runtime.clone();
+
+        RowGroupDecoder {
+            paths,
+            hive_partitions,
+            include_file_paths,
+            projected_arrow_fields,
+            row_index,
+            physical_predicate,
+            cpu_runtime,
+        }
+    }
+
+    fn init_projected_arrow_fields(&mut self) {
+        let reader_schema = self
+            .file_info
+            .reader_schema
+            .as_ref()
+            .unwrap()
+            .as_ref()
+            .unwrap_left()
+            .clone();
+
+        self.projected_arrow_fields =
+            if let Some(columns) = self.file_options.with_columns.as_deref() {
+                columns
+                    .iter()
+                    .map(|x| {
+                        // `index_of` on ArrowSchema is slow, so we use the polars native Schema,
+                        // but we need to remember to subtact the row index.
+                        let pos = self.file_info.schema.index_of(x.as_str()).unwrap()
+                            - (self.file_options.row_index.is_some() as usize);
+                        reader_schema.fields[pos].clone()
+                    })
+                    .collect()
+            } else {
+                Arc::from(reader_schema.fields.as_slice())
+            };
+
+        if self.verbose {
+            eprintln!(
+                "[ParquetSource]: {} columns to be projected from {} files",
+                self.projected_arrow_fields.len(),
+                self.paths.len(),
+            );
+        }
+    }
+
+    fn get_byte_source_builder(&self) -> DynByteSourceBuilder {
+        let byte_source_builder =
+            if is_cloud_url(self.paths[0].to_str().unwrap()) || config::force_async() {
+                DynByteSourceBuilder::ObjectStore
+            } else {
+                DynByteSourceBuilder::Mmap
+            };
+
+        if let Ok(v) = std::env::var("POLARS_BYTE_SOURCE") {
+            match v.as_str() {
+                "Mmap" => DynByteSourceBuilder::Mmap,
+                "ObjectStore" => DynByteSourceBuilder::ObjectStore,
+                "AsyncFile" => DynByteSourceBuilder::AsyncFile,
+                "CopyFromMmap" => DynByteSourceBuilder::CopyFromMmap,
```

Comment:
There was some consideration about not using mmap to avoid inadvertent page faulting / blocking I/O on the executor threads, so I added a few other options for how we read bytes that can be toggled. However, in my testing `Mmap` was generally the fastest, so I have set that as the default. The others performed the same as `ObjectStore`.

Reply:
I ran some benchmarks with the following:
* Pre-fetch methods:
  * No prefetch
  * L2 pre-fetch (based on `MemSlice::prefetch`, which IIUC seems to issue pre-fetch instructions for the entire slice)
  * madvise MADV_WILLNEED
  * madvise MADV_POPULATE_READ (linux only)
  * madvise MADV_SEQUENTIAL
* Reads (on a file with 50M rows, 51 columns, ~250k row group size):
  * Full projection
  * Project 30 / 51 columns
* Systems:
  * MacOS
  * Linux VM under MacOS

The performance was essentially identical in all but one case - when performing a full projection in Linux, using MADV_SEQUENTIAL significantly reduced the worst-case runtimes:

```python
[ParquetSource] Memory prefetch function: no_prefetch
python .env/y.py 2> /dev/null > /dev/null  22.05s user 61.95s system 647% cpu 12.981 total
python .env/y.py 2> /dev/null > /dev/null  22.93s user 91.01s system 605% cpu 18.825 total
python .env/y.py 2> /dev/null > /dev/null  27.93s user 55.58s system 655% cpu 12.747 total
python .env/y.py 2> /dev/null > /dev/null  23.19s user 91.12s system 609% cpu 18.763 total
python .env/y.py 2> /dev/null > /dev/null  28.23s user 57.20s system 669% cpu 12.763 total
[ParquetSource] Memory prefetch function: prefetch_l2
python .env/y.py 2> /dev/null > /dev/null  29.03s user 57.29s system 664% cpu 12.989 total
python .env/y.py 2> /dev/null > /dev/null  23.10s user 91.97s system 612% cpu 18.781 total
python .env/y.py 2> /dev/null > /dev/null  28.68s user 55.70s system 667% cpu 12.636 total
python .env/y.py 2> /dev/null > /dev/null  22.56s user 91.10s system 603% cpu 18.824 total
python .env/y.py 2> /dev/null > /dev/null  29.23s user 59.48s system 666% cpu 13.308 total
[ParquetSource] Memory prefetch function: madvise_willneed
python .env/y.py 2> /dev/null > /dev/null  29.28s user 61.62s system 669% cpu 13.578 total
python .env/y.py 2> /dev/null > /dev/null  23.31s user 85.63s system 613% cpu 17.762 total
python .env/y.py 2> /dev/null > /dev/null  28.40s user 60.41s system 669% cpu 13.261 total
python .env/y.py 2> /dev/null > /dev/null  23.12s user 87.48s system 621% cpu 17.807 total
python .env/y.py 2> /dev/null > /dev/null  28.10s user 57.67s system 660% cpu 12.988 total
[ParquetSource] Memory prefetch function: madvise_populate_read
python .env/y.py 2> /dev/null > /dev/null  30.49s user 61.92s system 657% cpu 14.050 total
python .env/y.py 2> /dev/null > /dev/null  22.89s user 91.27s system 585% cpu 19.494 total
python .env/y.py 2> /dev/null > /dev/null  28.98s user 56.33s system 665% cpu 12.818 total
python .env/y.py 2> /dev/null > /dev/null  22.94s user 92.60s system 587% cpu 19.674 total
python .env/y.py 2> /dev/null > /dev/null  28.89s user 57.12s system 653% cpu 13.167 total
[ParquetSource] Memory prefetch function: madvise_sequential
python .env/y.py 2> /dev/null > /dev/null  23.81s user 60.52s system 669% cpu 12.593 total
python .env/y.py 2> /dev/null > /dev/null  23.01s user 59.47s system 667% cpu 12.361 total
python .env/y.py 2> /dev/null > /dev/null  23.12s user 57.99s system 671% cpu 12.072 total
python .env/y.py 2> /dev/null > /dev/null  22.99s user 57.09s system 668% cpu 11.988 total
python .env/y.py 2> /dev/null > /dev/null  23.04s user 57.49s system 671% cpu 11.998 total
```

For all other runs, the difference was negligible, e.g. on ARM MacOS:
```python
[ParquetSource] Memory prefetch function: no_prefetch
python .env/y.py 2> /dev/null > /dev/null  16.49s user 7.25s system 456% cpu 5.201 total
python .env/y.py 2> /dev/null > /dev/null  16.49s user 7.14s system 458% cpu 5.155 total
python .env/y.py 2> /dev/null > /dev/null  16.58s user 6.93s system 467% cpu 5.030 total
python .env/y.py 2> /dev/null > /dev/null  16.54s user 7.20s system 482% cpu 4.924 total
python .env/y.py 2> /dev/null > /dev/null  16.54s user 6.90s system 466% cpu 5.026 total
[ParquetSource] Memory prefetch function: prefetch_l2
python .env/y.py 2> /dev/null > /dev/null  16.54s user 7.04s system 471% cpu 5.001 total
python .env/y.py 2> /dev/null > /dev/null  16.61s user 7.20s system 473% cpu 5.029 total
python .env/y.py 2> /dev/null > /dev/null  16.51s user 7.02s system 469% cpu 5.008 total
python .env/y.py 2> /dev/null > /dev/null  16.60s user 6.92s system 468% cpu 5.017 total
python .env/y.py 2> /dev/null > /dev/null  16.67s user 7.19s system 470% cpu 5.074 total
[ParquetSource] Memory prefetch function: madvise_willneed
python .env/y.py 2> /dev/null > /dev/null  16.47s user 7.20s system 468% cpu 5.052 total
python .env/y.py 2> /dev/null > /dev/null  16.54s user 6.88s system 469% cpu 4.994 total
python .env/y.py 2> /dev/null > /dev/null  16.61s user 7.15s system 475% cpu 4.995 total
python .env/y.py 2> /dev/null > /dev/null  16.54s user 7.00s system 472% cpu 4.984 total
python .env/y.py 2> /dev/null > /dev/null  16.59s user 7.03s system 476% cpu 4.961 total
[ParquetSource] Memory prefetch function: madvise_sequential
python .env/y.py 2> /dev/null > /dev/null  16.17s user 7.68s system 424% cpu 5.612 total
python .env/y.py 2> /dev/null > /dev/null  16.39s user 7.43s system 460% cpu 5.171 total
python .env/y.py 2> /dev/null > /dev/null  16.25s user 7.62s system 431% cpu 5.532 total
python .env/y.py 2> /dev/null > /dev/null  16.87s user 7.35s system 455% cpu 5.311 total
python .env/y.py 2> /dev/null > /dev/null  16.23s user 7.52s system 423% cpu 5.603 total
```

Reply:
@nameexhaustion I've found in the past that ARM MacOS quite aggressively caches disk data in memory. Can you try re-running the benchmark on MacOS with `sync && sudo purge` in between invocations?

For linux `madvise_sequential` sounds good 👍

Reply:
On MacOS with purging, it seems the sequential hint gives the slowest performance - 

```python
[ParquetSource] Memory prefetch function: no_prefetch
python .env/y.py 2> /dev/null > /dev/null  14.64s user 6.95s system 301% cpu 7.166 total
python .env/y.py 2> /dev/null > /dev/null  15.02s user 7.08s system 314% cpu 7.030 total
python .env/y.py 2> /dev/null > /dev/null  15.02s user 6.97s system 316% cpu 6.955 total
python .env/y.py 2> /dev/null > /dev/null  15.06s user 6.87s system 313% cpu 6.996 total
python .env/y.py 2> /dev/null > /dev/null  15.12s user 6.95s system 319% cpu 6.915 total
[ParquetSource] Memory prefetch function: prefetch_l2
python .env/y.py 2> /dev/null > /dev/null  14.79s user 6.85s system 318% cpu 6.796 total
python .env/y.py 2> /dev/null > /dev/null  15.09s user 6.89s system 317% cpu 6.927 total
python .env/y.py 2> /dev/null > /dev/null  15.08s user 6.91s system 320% cpu 6.849 total
python .env/y.py 2> /dev/null > /dev/null  14.54s user 6.84s system 313% cpu 6.812 total
python .env/y.py 2> /dev/null > /dev/null  14.96s user 6.93s system 319% cpu 6.854 total
[ParquetSource] Memory prefetch function: madvise_willneed
python .env/y.py 2> /dev/null > /dev/null  15.73s user 7.66s system 316% cpu 7.385 total
python .env/y.py 2> /dev/null > /dev/null  15.44s user 7.54s system 314% cpu 7.317 total
python .env/y.py 2> /dev/null > /dev/null  15.47s user 7.30s system 311% cpu 7.320 total
python .env/y.py 2> /dev/null > /dev/null  15.79s user 7.96s system 308% cpu 7.698 total
python .env/y.py 2> /dev/null > /dev/null  15.93s user 7.81s system 308% cpu 7.705 total
[ParquetSource] Memory prefetch function: madvise_sequential
python .env/y.py 2> /dev/null > /dev/null  14.31s user 7.63s system 255% cpu 8.581 total
python .env/y.py 2> /dev/null > /dev/null  14.52s user 7.80s system 260% cpu 8.571 total
python .env/y.py 2> /dev/null > /dev/null  14.14s user 7.67s system 252% cpu 8.627 total
python .env/y.py 2> /dev/null > /dev/null  14.77s user 7.51s system 262% cpu 8.471 total
python .env/y.py 2> /dev/null > /dev/null  14.18s user 7.68s system 255% cpu 8.550 total
```

I think for MacOS we can leave it as no prefetch

Reply:
Alright let's do that for now, we can always revisit it later.